File: Domain/VersionItem.cs
using System;
using System.Runtime.Serialization;

namespace Domain
{
    [DataContract]
    public class VersionItem : IDocumentKey
    {
        [DataMember(Name = "id")]
        public string Id { get; set; }

        [DataMember(Name = "author_id")]
        public string AuthorId { get; set; }

        [DataMember(Name = "target_id")]
        public int TargetId { get; set; }

        [DataMember(Name = "target_type")]
        public string TargetType { get; set; }

        [DataMember(Name = "timestamp")]
        public DateTimeOffset Timestamp { get; set; }

        [DataMember(Name = "parent_id")]
        public string ParentId { get; set; }

        public string Key => $"{Type}::{Id}";

        [DataMember(Name = "type")]
        public string Type { get; } = "version_item";
    }
}

File: DatabaseTypes/VersionItem.fs
namespace DatabaseTypes

open System
open System.Runtime.Serialization

[<DataContract>]
type VersionItem =
    {
        [<field: DataMember(Name = "id")>]
        Id : string
        [<field: DataMember(Name = "author_id")>]
        AuthorId : string
        [<field: DataMember(Name = "target_id")>]
        TargetId : string
        [<field: DataMember(Name = "target_type")>]
        TargetType : string
        [<field: DataMember(Name = "timestamp")>]
        Timestamp : DateTimeOffset
        [<field: DataMember(Name = "parent_id")>]
        ParentId : string
    }

    static member TypeName = "version_item"
    static member CreateDocumentKey(id: string): DocumentKey =
        DocumentKey.Create(id, VersionItem.TypeName)
    member private this.DocKey = VersionItem.CreateDocumentKey(this.Id)

    interface IDocumentKey with
        [<DataMember(Name = "type")>]
        member this.Type
            with get() = this.DocKey.Type
        member this.Key
            with get() = this.DocKey.Key
